frontend/catalyst/ag_primitives.py

```diff
@@ -318,7 +327,87 @@ def for_stmt(
     # Sometimes we unpack the results of nested tracing scopes so that the user doesn't have to
     # manipulate tuples when they don't expect it. Ensure set_state receives a tuple regardless.
     if not isinstance(results, tuple):
-        results = (results,)
+        # TODO: Do we have a use-case for this brnach?
+        results = (results,)  # pragma: nocover
+    set_state(results)
+
+
+def _call_catalyst_while(loop_test, loop_body, get_state, set_state, _nonlocals, _symbol_names):
+    """Dispatch to a Catalyst implementation of while loops."""
+
+    if _emulate_fallback_errors:
+        raise AutoGraphError("Emulated autograph fallback errror")
+
+    def _test(state):
+        old = get_state()
+        set_state(state)
+        res = loop_test()
+        set_state(old)
+        return res
+
+    @catalyst.while_loop(_test)
+    def _functional_while(iter_args):
+        set_state(iter_args)
+        loop_body()
+        return get_state()
+
+    iter_inits = get_state()
+    iter_results = _functional_while(iter_inits)
+    return iter_results
+
+
+def _call_python_while(loop_test, loop_body, get_state, _set_state, _nonlocals, _symbol_names):
+    """Fallback to a Python implementation of while loops."""
+
+    while loop_test():
+        loop_body()
+
+    return get_state()
+
+
+def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_names):
+    """An implementation of the AutoGraph 'while ..' statement. The interface is defined by
+    AutoGraph, here we merely provide an implementation of it in terms of Catalyst primitives."""
+
+    fallback = False
+
+    try:
+        results = _call_catalyst_while(
+            loop_test, loop_body, get_state, set_state, nonlocals, symbol_names
+        )
+
+    except Exception as e:  # pylint: disable=broad-exception-caught
+        if catalyst.autograph_strict_conversion:
+            raise e
+
+        # pylint: disable=import-outside-toplevel
+        import inspect
+        import textwrap
+
+        fallback = True
+
+        while_loop_info = get_source_code_info(inspect.stack()[1])
+
+        if not catalyst.autograph_ignore_fallbacks:
+            warnings.warn(
+                f"Tracing of an AutoGraph converted while loop failed with an exception:\n"
+                f"  {type(e).__name__}:{textwrap.indent(str(e), '    ')}\n"
+                f"\n"
+                f"The error ocurred within the body of the following for loop statement:\n"
+                f"{while_loop_info}"
+                f"\n"
+                f"To understand different types of JAX tracing errors, please refer to the "
+                f"guide at: https://jax.readthedocs.io/en/latest/errors.html\n"
+                f"\n"
+                f"If you did not intend for the conversion to happen, you may safely ignore "
+                f"this warning."
+            )
```

Comment:
Do you think the warning is actually to users, at least in some instances?

Reply:
Basically, the warning exists in the for loop so users can fix their code to make the conversion succeed (see for loop test cases). If we can't think of any such case for the while loop, we shouldn't raise a warning because it is not actionable for the used, and they are more likely to just silence all warnings.

Reply:
I think we need to keep the behavior observable anyways. Inability to insert `@cond_while` would almost certainly lead to code expansion which is something we should want to communicate. It might be useful for us as well, e.g. in issue reports. The contents of the warning might be different though. Let me know what do you think.

Reply:
Removing the fallback does not seem sensible if we're trying to build a robust system.

> Inability to insert @while_loop would almost certainly lead to code expansion which is something we should want to communicate

Not necessarily, there is going to be loops that are not convertible (and maybe weren't intended to be). The default approach is to perform conversion on a best effort basis, allowing unconvertible constructs to execute in Python instead. This leads to an overall good experience that "just works" (as best as we can achieve this). Always raising warnings whenever this happens would be a pain point for users.

I agree that logging this type of information could be important however, for this purpose autograph has a global debug flag with different levels. We should think about this, but it would be a separate task I think.

Reply:
> Always raising warnings whenever this happens would be a pain point for users.

Not sure about the amount of pain (You might be right, I just don't know. Isn't it a painful to understand the cause of source expansion after hours of debugging?). I still think that removing this warning would make important information invisible or less visible, depending on the autograph logging capabilities you mentioned below. 

> I agree that logging this type of information could be important however, for this purpose autograph has a global debug flag with different levels. We should think about this, but it would be a separate task I think.

I also relied on this warning in testing. Without it I think I need to replace some tests with lit-tests on JAX sources. It might be useful by itself.

I still don't see why the absence of this warning is consistent with what we have for for-loops (and our logical-and-or approach looks to me also different from both these cases as well). I think it is a management/UI decision in the end, so please let me know what implementation do you want to see here.

Reply:
> Not sure about the amount of pain (You might be right, but there are sociological questions for which I don't know answers: Isn't it a painful to understand the cause of source expansion after hours of debugging? What scenario would lead to a more negative user reaction?) I only think that removing this warning might make important information invisible.

Well, I can only tell you that this is Josh's policy regarding warnings, and I'm inclined to agree with it. Warnings that can't be acted upon aren't all that useful.

> I still don't see why the absence of this warning is consistent with what we have for for-loops (and our logical-and-or approach looks to me also different from both these cases as well). I think it is a management/UI decision in the end, so if you are not convinced please let me know and I will remove it.

There is a difference. In the for loop case there are *concrete instances* that we know will fail conversion but are fixable by the user. Raising a warning then allows them to fix it.
There are are also instances where the user can't do anything to fix it and we raise the warning. This is bad, but we accept it because we can't differentiate between the fixable and unfixable case. To mitigate this, we added the configuration to allow users to silence the warnings.
Also note that there are [fallbacks in the for loop case](https://github.com/PennyLaneAI/catalyst/blob/445b15881d71fb454b05782b5d55d8526272a82d/frontend/catalyst/ag_primitives.py#L247-L262) that we don't raise a warning for, precisely because these aren't instances the user could fix, and were probably meant as Python loops in the first place.
In the case of while loop, I don't know any instance that fails conversion that the user could correct.

In short, we use warnings only sparingly and when there is immediate action the user can take to resolve the warning.

What you propose has value in a debug or verbose mode, and we should look into that, but using warnings for it doesn't match the experience we want users to have.

Reply:
As discussed, I have removed the warning.